src/main/java/net/snowflake/ingest/streaming/internal/CustomCompressionCodec.java

```diff
@@ -0,0 +1,118 @@
+package net.snowflake.ingest.streaming.internal;
+
+import org.apache.arrow.compression.Lz4CompressionCodec;
+import org.apache.arrow.compression.ZstdCompressionCodec;
+import org.apache.arrow.memory.ArrowBuf;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.util.MemoryUtil;
+import org.apache.arrow.vector.compression.CompressionCodec;
+import org.apache.arrow.vector.compression.CompressionUtil;
+
+/**
+ * This is custom wrapper on top of {@link CompressionCodec} implementations.
+ *
+ * <p>The main stream {@link CompressionCodec} implementations are based on {@link
+ * org.apache.arrow.vector.compression.AbstractCompressionCodec}. This implementation is similar to
+ * it with some tweaks.
+ *
+ * <p>The reason to have this custom implementation is to fix certain issues with mainstream codecs
+ * that lead to buffer reference counting problems due to creation of new compressed buffers and
+ * releasing uncompressed. The release of uncompressed buffer from the original vectors to write
+ * into file results in double release when the original vector gets released as usual.
+ *
+ * <p>The custom code extraction is not optimal in a long term and there is a TODO: to follow up on
+ * this.
+ */
+public class CustomCompressionCodec implements CompressionCodec {
+  private final CompressionCodec actualCodec;
+
+  public CustomCompressionCodec(CompressionUtil.CodecType codecType) {
+    switch (codecType) {
```

Comment:
I would suggest we remove all the other compression options if all we support now is ZSTD, we should add them whenever it's needed

Reply:
I don't think we use NO_COMPRESSION neither, should we remove it? Or you believe that we will use it later?